docs/namespace_config_spec.md

```diff
@@ -0,0 +1,281 @@
+declarations.
+
+```ebnf
+Config          = [ ClassDecl ] .
+ClassDecl       = "type" identifier "implements" "Namespace" "{" ClassSpec "}" .
+ClassSpec       = MetadataDefn | [ RelationDecls ] | [ PermissionDefns] 
+```
+
+#### Example
+
+The following declares a type *User* that is never the object in any
+relation. It can, however, be the subject of a relation, the relation being
+declared on some other type.
+
+```typescript
+class User implements Namespace { metadata = { id: 1 } }
+```
+
+### Relation declaration
+
+Inside the class declaration, relations are declared. `RelationName` must be
+unique identifiers and are used as the relation strings in the tuples. The
+`TypeName` must be the name of a `type` that is defined above or below (in
+TypeScript: a class that implements `Namespace`).
+
+```ebnf
+RelationDecls   = "relations" ":" "{" { RelationName ":" TypeName "[]" } "}" .
+RelationName    = literal .
+TypeName        = literal .
+```
+
+Note that all relations are defined as array types `T[]`, because there are
+naturally only many-to-many relations in Keto.
+
+The following declares a type *Document* has three relations: *owners* and
+*viewers*, both of which have *users* as subjects. Additionally, the relation
+*parent* has type *Document*.
+```typescript
+type User { metadata = { id: 1 } }
+
+type Document {
+  metadata = { id: 2 }
+  relations = {
+    parent: Document[]
+    owners: User[]
+    viewers: User[]
+  }
+}
+```
+
+### Permission definition
+
+Again inside the class declarations, the permissions are defined as functions
+that thake a `ctx` of type `Context` and evaluate to a boolean `true` or
+`false`.
+
+```ebnf
+PermissionDefns = "permissions" "=" "{" Permission [ "," Permission ] "}" .
+Permission      = PermissionSign "=>" PermissionBody .
+PermissionSign  = PermissionName ":" "(" "ctx" [ ":" "Context" ] ")" [ ":" "boolean" ] .
+PermissionName  = identifier .
+```
+
+The `ctx` object is a fixed parameter that contains the `subject` for which the
+permission check should be conducted:
+```typescript
+ctx = {subject: "some_user_id"}
+```
+In the future, more context fields could be added here, e.g., the IP range or
+geolocation, the time of day, or the security level of the device making the
+request.
+
+
+```ebnf
+PermissionBody  = PermissionCheck | { Operator PermissionCheck } .
+Operator        = "||" | "&&" .
+PermissionCheck = TransitiveCheck | IncludesCheck .
+```
+
+The body of a permission check is either one of
+
+* a `IncludesCheck`, a check that something is in a set, e.g., `this.relations.viewers.includes(ctx.subject)`:
+```ebnf
+IncludesCheck   = "this" "." "relations" "." RelationName "." "includes" "(" "ctx" "." "subject" ")" .
+```
+* a `TranstitiveCheck`, a call to a permission on a relation, e.g., `this.relations.parents.some(p => p.permissions.view(ctx))`:
+```ebnf
+TransitiveCheck = "this" "." "relations" "." RelationName "." "some" "(" Var "=>" PermissionCall ")" .
```

Comment:
We decided to leave this out of the first version, since that semantic does not translate to Zanzibar concepts yet (Zanzibar has no set intersection for rewrites). For top-level intersection / every / and, you can use `&&`.